[WIP] Issue #61 Fixed linting issues

kishansagathiya opened the pull request:
Fixes #61

jaseemabid commented on internal/openshift/user_channel_map.go:
```diff
@@ -1,35 +1,40 @@
 }
 
-// A type-safe and concurrent map storing channels for model.User keyed against the user namespace.
+// NewUserChannelMap creates a new instance of UserChannelMap
```
@kishansagathiya We lost some information here, didn't we?

kishansagathiya replied:
transferred here https://github.com/fabric8-services/fabric8-jenkins-idler/pull/142/files/534eda6b77bfb7eaae118b0c15948f6f2e247f0a#diff-86cb10d5a84e282eda5bce50248f298cR9

jaseemabid commented on internal/openshift/user_channel_map.go:
```diff
@@ -1,35 +1,40 @@
 	return result, ok
 }
 
+// Delete deletes the channe, of model.User from UserChannelMap given its key
```
Typo

jaseemabid commented on internal/tenant/model.go:
```diff
@@ -2,35 +2,44 @@ package tenant
 
-type TenantInfoList struct {
-	Data []TenantInfoData
+// InfoList defines list of informations about a tenant
```
This is indeed the naming convention as per go best practices, but I'm surprised to see you rename just one of these instead of doing it everywhere. Progress indeed, so LGTM.

kishansagathiya commented:
@jaseemabid this is so far WIP (work in progress). It is just to show the progress. Mostly I will be adding code to this and won't have to modify much, but ideally, you should review once it is ready or you will waste your time in looking at something which is not fully done.

kishansagathiya commented:
@hferentschik 

> Also your commit message does not pass the expected format. It should be something like:
"Issue #61 Fixing golint linting issues"

Does it matter as it will anyway be squashed before being merged ( I hope that's the case)
Will follow this practice in coming commits. Was not aware of it.

hferentschik commented:
> Does it matter as it will anyway be squashed before being merged 

Whether it get squashed is a different question. At some stage I expect this pull request to pass the build. It won't unless you have commit messages which don't pass the validation. This is regardless of whether you have one or more commits.

Whether one or multiple commits make sense will depend on the task/issue. In some cases it makes sense to group commits. For this particular issue, I'd also would only expect a single commit. 

It is your responsibility to create a pull request which is "mergeable". It should be rebased on master and pass all tests. In some cases, I might do the rebase or add an amending commit, but usually you as author of the change are responsible to getting this into a state where it can be merged/applied directly.

I'd would just keep pushing a single commit to this pull request.

hferentschik commented:
BTW, you realise that you are breaking unit tests as well. You will need to update the test expectations as well. We also can have a discussion on whether we want to pass golint in its default config, or whether we want to customise a bit, eg the setting that a error message should be lower case and no punctuation. On this particular one, I could swing either way.

kishansagathiya commented:
@hferentschik Oh I just saw it now that this commit message rule is added in `make all`. Now I know why you asked for it. 
Btw you will have to keep in mind that I was not aware of these conventions :smiley:
Again this is Work In Progress, not ready for review.

kishansagathiya commented:
@hferentschik 
Few linting issues are still remaining. Most of them are about adding comments to the structs. I am not very clear on what they are and why do we have them. So I would need help from you for proper descriptions of these structs.

hferentschik commented:
> Btw you will have to keep in mind that I was not aware of these conventions 

No harm done at all. I was just pointing it out. There is a comment in the CONTRIBUTING page https://github.com/fabric8-services/fabric8-jenkins-idler/blob/master/CONTRIBUTING.md

I guessed that  you might have wondered why things failed.

> Again this is Work In Progress, not ready for review.

Fully understood. For this reason, I did not look at the actual changes. I just wanted to make sure you add the 'lint' target to 'all'. This way it becomes obvious when you are close to ready, since it will pass the build. For now, I would need to pull your pull request and run 'lint' locally.

hferentschik commented:
> Few linting issues are still remaining. Most of them are about adding comments to the structs. I am not very clear on what they are and why do we have them. So I would need help from you for proper descriptions of these structs.

Ok, no worries. Let me take a look.

kishansagathiya commented:
@hferentschik 
> For now, I would need to pull your pull request and run 'lint' locally.

`lint` is added to `make all` in latest commit

hferentschik commented:
> lint is added to make all in latest commit

Sweet, thanks

hferentschik commented:
So what these objects are, are the struct form of the various JSON objects which you receive when you watch OpenShift for changes. model.Object is what goes over the wire and it gets unmarshalled into these structs. We only define the fields in the struct we are interested in. Golang ignores other JSON fields. The full definition of the JSON is here https://docs.openshift.com/online/rest_api/oapi/v1.Build.html and https://docs.openshift.com/online/rest_api/oapi/v1.DeploymentConfig.html

If you compare the structs with the format specified in the link, you should see that they match (just that the structs do not cover everything)

Effectively, the Idler does watch "builds" (OpenShift.WatchBuilds) and "deploymentconfigs"(OpenShift.WatchDeploymentConfigs)

You should be able to figure out now what the various structs mean, right? I suggest to add the links above as reference. Does that help?

hferentschik commented on Makefile:
```diff
@@ -26,7 +26,7 @@ __check_defined = \
       $(error Undefined $1$(if $2, ($2))))
 
-all: tools build test fmtcheck validate_commits vet image ## Compiles binary and runs format and style checks
+all: tools build test fmtcheck validate_commits vet image lint ## Compiles binary and runs format and style checks
```
+1

hferentschik replied:
Let's add something like

    LINT_PACKAGES = $(shell echo $(PACKAGES) | sed -e 's@github.com/fabric8-services/fabric8-jenkins-idler/internal/configuration@@')

under the PACKAGES definition above. And then use this variable in 'lint'

hferentschik replied:
The rub is really that golint does not provide any mechanism to prevent a check. These are really suggestions as mentioned in the README. On one side I really would like to be as close as possible to the suggestions, but it comes at a cost. 

Problem with not failing a build when there are violations is that they will eventually sneak in. For now, I'd go with the change to the Makefile. I find most of the changes made as part of this pull request reasonable otherwise. One can discuss changes in policies and enforcement of rules when another potential conflict arises. @
kishansagathiya, does tha make sense?

hferentschik replied:
BTW, please make 'all' phony as well. Just a goos measure

kishansagathiya replied:
Makes sense. 
Since we lint would apply only on listed package, we would be able to add ALL_CAPS variable names as we have in config

kishansagathiya replied:
Doesn't it make sense to keep all targets as phony?

hferentschik replied:
no, $(BUILD_DIR), $(BUILD_DIR)/$(REGISTRY_IMAGE), image, tools are all file based and are not phony

hferentschik commented on cmd/fabric8-jenkins-idler/idler.go:
```diff
@@ -52,33 +53,33 @@ func (idler *Idler) Run() {
 	)
 
-	startWorkers(&wg, ctx, cancel, openShift, controller, idler.config.GetDebugMode())
+	startWorkers(ctx, &wg, cancel, openShift, controller, idler.config.GetDebugMode())
```
ooi, is there a rule that says the context should come first?

kishansagathiya replied:
Yup

hferentschik commented on internal/api/api.go:
```diff
@@ -36,15 +37,15 @@ type IdlerAPI interface {
 type idler struct {
 	openShiftClient client.OpenShiftClient
-	controller      openshift.Controller
+	controller      openshift.ControllerI
```
ControllerI? Did you rename the Controller interface to ControllerI? I am not really happy with this one. Let's stick to Controller as interface and try finding another name for what used to be OpenShiftController. We can keep 'Controller' as interface name, since this is what's actually exposed everywhere. And the actual implementation could be "controller". It does not even have to be exported. Or controllerImpl.

hferentschik commented on internal/condition/condition.go:
```diff
@@ -14,10 +14,12 @@ type Condition interface {
 	Eval(object interface{}) (bool, error)
 }
 
+// Conditions defines list of Condition instances by their names
```
It is not a list, but a map. You cannot define "a list by their names". You can key a set  of conditions against their name.

hferentschik commented on internal/condition/deployment_config_condition.go:
```diff
@@ -1,17 +1,18 @@
 )
 
-// DCCondition covers changes to Jenkins DeploymentConfigs
+// DeploymentConfigCondition covers changes to Jenkins DeploymentConfigs
```
that is not correct per se. It does not have to be Jenkins. It is a DeploymentConfiguration in general.

hferentschik commented on internal/configuration/configuration.go:
```diff
@@ -22,31 +23,33 @@ var (
 
 func init() {
 	// service URLs for required services
-	settings["GetOpenShiftURL"] = Setting{"JC_OPENSHIFT_API_URL", "", []func(interface{}, string) error{util.IsURL}}
```
!? We cannot do that. That defeats the purpose here. This are the actual names of the envirnoment variables.

hferentschik replied:
I am pretty sure you broke things with this change.

hferentschik commented on internal/configuration/configuration_test.go:
```diff
@@ -2,24 +2,25 @@ package configuration
-	JC_MAX_RETRIES           string
-	JC_CHECK_INTERVAL        string
-	errors                   []string
+	JCOpenShiftAPIURL    string
```
-1

hferentschik commented on internal/openshift/controller.go:
```diff
@@ -24,14 +25,15 @@ const (
 
-// Controller defines the interface for watching the OpenShift cluster for changes.
-type Controller interface {
+// ControllerI defines the interface for watching the OpenShift cluster for changes.
```
-1

kishansagathiya commented:
@hferentschik I have addressed all of your reviews. Few linting issues are still left. I am going to spend some time reading the documentation for that.

kishansagathiya commented:
@hferentschik  Still can't figure out what to put here. Except this everything else is done.
```
 type Endpoint struct {
  	Metadata Metadata `json:"metadata"`
  }		  
```
https://github.com/fabric8-services/fabric8-jenkins-idler/pull/142/files#diff-27d39bd87d6eeb89b1eb746eec2caed9R65

kishansagathiya commented on internal/condition/deployment_config_condition.go:
```diff
@@ -1,17 +1,18 @@
 }
 
+// NewDCCondition creates a new instance of DeploymentConfigCondition
 func NewDCCondition(idleAfter time.Duration) Condition {
```
Wouldn't NewDeploymentConfigCondition be more appropriate?

hferentschik replied:
+1 Works for me

kishansagathiya commented on internal/model/openshift_objects.go:
```diff
@@ -10,39 +10,50 @@ import (
 }
 
+// DCObject is DeploymentConfig Object
 type DCObject struct {
```
This one too
DeploymentConfigObject ?

hferentschik replied:
+1

hferentschik replied:
You need to check whether this rename would break anything in terms of JSON marshalling though. I don't think so, but worth double checking

kishansagathiya commented on internal/model/openshift_objects.go:
```diff
@@ -55,18 +66,22 @@ type Endpoint struct {
 }
 
+// DCStatus : DeploymentConfig Status
 type DCStatus struct {
```
DeploymentConfigStatus ?

hferentschik commented:
Personally, I would not mind if you squash this now with a proper summarizing commit message. Get it into a state you would feel happy with applying it.

hferentschik commented on Makefile:
```diff
@@ -6,6 +6,7 @@ IMAGE_TAG ?= $(shell git rev-parse --short HEAD)
 
 BUILD_DIR = out
 PACKAGES = $(shell go list ./...)
+LINT_PACKAGES = $(shell echo $(PACKAGES) | sed -e 's@github.com/fabric8-services/fabric8-jenkins-idler/internal/configuration@@')
```
+1, could be improved further down the road if we have more than one package, file to exlude. For now it should do.

hferentschik commented on cmd/fabric8-jenkins-idler/idler.go:
```diff
@@ -52,33 +53,33 @@ func (idler *Idler) Run() {
 
 	// Create Idler controller
-	controller := openshift.NewOpenShiftController(
+	controller := openshift.NewControllerImpl(
```
Do you really need 'Impl'?

kishansagathiya replied:
Nope, but if the struct is `ControllerImpl` then it makes sense to have a function named `NewControllerImpl`

hferentschik replied:
Let me rephrase, why do you need to call it ControllerImpl? Don't you get away with 'controller'?

hferentschik commented on internal/condition/build_condition.go:
```diff
@@ -22,7 +23,7 @@ func NewBuildCondition(idleAfter time.Duration) Condition {
 	u, ok := object.(model.User)
 	if !ok {
-		return false, errors.New(fmt.Sprintf("%T is not of type User", object))
+		return false, fmt.Errorf("%T is not of type User", object)
```
Do you know the motivation for Errorf over errors.New?

kishansagathiya replied:
It seems like `fmt.Errorf(.....)` is same as `errors.New(fmt.Sprintf(......))`
From golang documentation 
fmt.Errorf() :  It formats a string according to Printf's rules and returns it as an error created by errors.New.
https://blog.golang.org/error-handling-and-go

hferentschik replied:
Sure the net effect is exactly the same, otherwise, I would be concerned. The question is why "Effective Go" (which is the foundation for this linting) prefers one over the other.

hferentschik commented on internal/condition/condition.go:
```diff
@@ -14,10 +14,12 @@ type Condition interface {
 	Eval(object interface{}) (bool, error)
 }
 
+// Conditions defines map of Condition instances by their names
```
+1

hferentschik commented on internal/configuration/configuration.go:
```diff
@@ -41,12 +42,14 @@ func init() {
 	validations  []func(interface{}, string) error
 }
 
+// Configuration has methods to get all configuration details of the current instance of the Jenkins Idler
```
"Configuration defines the configuration options of the Idler"

hferentschik commented on internal/configuration/configuration_test.go:
```diff
@@ -28,14 +29,14 @@ func Test_configuration_settings(t *testing.T) {
-		{"http://localhost", "", "http://localhost", "http://localhost", "token-1", "token-2", "10", "15", "15", []string{"Value for JC_JENKINS_PROXY_API_URL needs to be a valid URL."}},
-		{"http://localhost", "http://localhost", "", "http://localhost", "token-1", "token-2", "10", "15", "15", []string{"Value for JC_F8TENANT_API_URL needs to be a valid URL."}},
-		{"http://localhost", "http://localhost", "http://localhost", "", "token-1", "token-2", "10", "15", "15", []string{"Value for JC_TOGGLE_API_URL needs to be a valid URL."}},
+		{"", "http://localhost", "http://localhost", "http://localhost", "token-1", "token-2", "10", "15", "15", []string{"value for JC_OPENSHIFT_API_URL needs to be a valid URL"}},
```
you just changed to start with lower case and no punctuation, right?

kishansagathiya replied:
yup

hferentschik replied:
Good

hferentschik commented on internal/configuration/configuration_test.go:
```diff
@@ -103,7 +104,8 @@ func getConfigValueForEnvKey(v *config, key string) (string, bool) {
-		return "", false
 	}
+
+	return "", false
```
+1

hferentschik commented on internal/idler/user_idler.go:
```diff
@@ -35,6 +37,7 @@ type UserIdler struct {
 	features        toggles.Features
 }
 
+// NewUserIdler creates an instance of UserIdler
```
You could be a bit more specific. You just state the obvious. I am not too fuzzed atm, but a good godocs would describe the parameters as well and give more info about the return type, for example that a pointer to a UserIdler is returned

hferentschik commented on internal/model/openshift_objects.go:
```diff
@@ -75,34 +90,41 @@ type DCStatus struct {
 	Strategy Strategy
 }
 
+// Strategy describes the build strategy used to execute the build
```
+1 Nice with the links

hferentschik commented on internal/model/user.go:
```diff
@@ -32,22 +34,26 @@ func NewUser(id string, name string) User {
 	}
 }
 
+// HasActiveBuilds checks if current user has any active builds
```
And what does it return?

"HasActiveBuilds checks if current user has any active builds. If so true is returned, otherwise false"

hferentschik commented on internal/model/user.go:
```diff
@@ -32,22 +34,26 @@ func NewUser(id string, name string) User {
 	return len(u.ActiveBuild.Metadata.Name) > 0
 }
 
+// HasCompletedBuilds checks if current User has any completed/done builds
```
See  above

hferentschik commented on internal/openshift/controller.go:
```diff
@@ -31,7 +32,8 @@ type Controller interface {
 
-type OpenShiftController struct {
+// ControllerImpl watches OpenShift cluser for changes and implements Controller
+type ControllerImpl struct {
```
Do you really need 'Impl'?

kishansagathiya replied:
I have to change it from OpenShiftController to something else. Can't keep `Controller`, because that the interface. Other name that I can think of is `OSController`, but `ControllerImpl` seems more appropriate to me.

hferentschik commented on internal/testutils/mock/mock_config.go:
```diff
@@ -16,54 +18,69 @@ type Config struct {
 	return c.Debug
 }
 
+// GetFixedUuids returns a slice of fixed user uuids. If set, a custom Features implementation is instantiated
```
+1

hferentschik commented on internal/model/openshift_objects.go:
```diff
@@ -55,18 +66,22 @@ type Endpoint struct {
 	Metadata Metadata `json:"metadata"`
```
Endpoint is the how a service is getting accessed, its "endpoint" - https://docs.openshift.com/online/rest_api/api/v1.Endpoints.html

We are modifying (PATCH) the Jenkins service endpoint in order to idle Jenkins - https://github.com/fabric8-services/fabric8-jenkins-idler/blob/4625c2a9974567151aaa2706b46697a72359b5f2/internal/openshift/client/openshift_client.go#L87

kishansagathiya replied:
I see what you did. Very punny!!

hferentschik replied:
We are actually not just scaling down. We are using route based idling. OpenShift is not fully supporting this atm. They are still working on a more direct API. Atm, if you annotate a service like we do here, OpenShift will do route based idling, meaning, the pod is scaled down, but automatically scaled up when the route is accessed. We try to keep the user away from accessing the tenant Jenkins route directly, but in case he does, the Jenkins instance should still come up.

hferentschik commented:
@kishansagathiya, I hope this exercise gave you a bit of a tour around the code. My guess is that you touched all major parts by now. Anything which sticks out or you are wondering what it is about? 

If it is related to documentation, we can try to address it as part of this pull request, otherwise, please make a note and we have a chat some time this week where we can go through your questions.

hferentschik commented:
BTW, feel free to close this pull request and open a new one. After a certain amount of comments, things get unwieldy. In this case, it is often easier to close the pull request and create a new slate.

kishansagathiya commented:
I would prefer to work on the same PR. It keeps all the discussions at the same place. IMO it would be a good practice for all of our PRs.

hferentschik commented:
> I would prefer to work on the same PR.

The old pull request is still there. You can reference it from the new one. But as you wish.

> It keeps all the discussions at the same place.

Nothing gets lost here. It often gets easier in terms of reviewing though.

hferentschik commented:
You would close this pull request with a comments like "superseded by bla" and the reverse on the new pull request.

kishansagathiya commented:
Superseded by https://github.com/fabric8-services/fabric8-jenkins-idler/pull/144
